05_concurrency/homework/main.go

```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type Cache interface {
+	Set(key string, value interface{}, ttl time.Duration)
+	Get(key string) (interface{}, bool)
+	Delete(key string)
+}
+
+type CacheElement struct {
+	Value       interface{}
+	ErasureTime time.Time
+	TTL         time.Duration
+}
+
+type CacheMemory struct {
+	sync.Mutex
+	Memory map[string]CacheElement
+}
+
+func (cache *CacheMemory) Delete(key string) {
+	cache.Lock()
+	delete(cache.Memory, key)
+	cache.Unlock()
+}
+
+func (cache *CacheMemory) Get(key string) (interface{}, bool) {
+	cache.Lock()
+	value, ok := cache.Memory[key]
+	if !ok {
+		cache.Unlock()
+		return nil, false
+	}
+	//updating the time of erase
+	cache.Memory[key] = CacheElement{
+		Value:       value.Value,
+		ErasureTime: time.Now().Add(cache.Memory[key].TTL),
+		TTL:         cache.Memory[key].TTL,
+	}
+	cache.Unlock()
+	return value.Value, true
```

Comment:
Returned value will have previous expiration time.Updated will be stored in memory. I don't mind, but from user proprective it could be better to return updated value with new expiration time on Get.

I am not sure if using `defer cache.Unlock()` can critically decrease performance in this case but it can significantly increase readability.